core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageInputStreamTest.java

```diff
@@ -318,4 +323,112 @@ void toInputStream_executor() throws IOException {
+        while ((read = inputStream.read()) != -1) {
+            result.writeByte(read);
+        }
+
+        final int readableBytes = result.readableBytes();
+        final byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        result.release();
+        assertThat(actual).isEqualTo(expected);
+        assertThat(inputStream.available()).isZero();
+    }
+
+    @Test
+    void httpDataConverter_error_thrown() throws IOException {
+        final StreamMessage<Integer> streamMessage = StreamMessage.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+        final InputStream inputStream = streamMessage
+                .toInputStream(x -> {
+                    if (x == 6) {
+                        return Exceptions.throwUnsafely(new RuntimeException());
+                    }
+                    return HttpData.wrap(x.toString().getBytes());
+                });
+
+        final byte[] expected = ImmutableList.of(1, 2, 3, 4, 5)
+                                             .stream()
+                                             .map(x -> x.toString().getBytes())
+                                             .reduce(Bytes::concat).get();
+
+        final ByteBuf result = Unpooled.buffer();
+        int read;
+        while ((read = inputStream.read()) != -1) {
+            result.writeByte(read);
+        }
+
+        final int readableBytes = result.readableBytes();
+        final byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        result.release();
+        assertThat(actual).isEqualTo(expected);
+        assertThat(inputStream.available()).isZero();
+    }
+
+    @Test
+    void maybeRequest_when_not_enough_data() throws IOException {
+        final StreamMessage<String> streamMessage1 = StreamMessage.of("12", "34");
+        final StreamMessage<String> streamMessage2 = spy(StreamMessage.of("56", "78"));
+        final InputStream inputStream = StreamMessage
+                .concat(streamMessage1, streamMessage2)
+                .toInputStream(x -> HttpData.wrap(x.getBytes()));
+
+        final ByteBuf result = Unpooled.buffer();
+        for (int i = 0; i < 2; i++) {
+            result.writeByte(inputStream.read());
+        }
+
+        int readableBytes = result.readableBytes();
+        byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        assertThat(actual).isEqualTo(ImmutableList.of("12")
+                                                  .stream()
+                                                  .map(String::getBytes)
+                                                  .reduce(Bytes::concat).get());
+        assertThat(inputStream.available()).isEqualTo(0);
+        verify(streamMessage2, never()).subscribe(any(), any(), any());
```

Comment:
Could we test this without mocking?
We may use `StreamWriter.whenConsumed()` and `StreamMessage.demand()` to know whether a request is made.

Reply:
✅ updated (use `whenConsumed` instead of `mocking`)